File: frontend/src/api/decisions.ts
/**
 * Triage-decisions API client — ADR-0072 D3, issue #47 Part 2/frontend.
 *
 * Mirrors the 3 /decisions endpoints from
 * packages/firewatch-api/src/firewatch_api/routes/decisions.py:
 *
 *   POST   /decisions            — record a decision (server computes the snapshot)
 *   GET    /decisions            — cursor-paginated history (ADR-0029 D2)
 *   DELETE /decisions/{id}       — soft-revoke (undo; audit row survives)
 *
 * SECURITY:
 *   - actor_ip / rule_name / note are bounded and pattern-validated server-side;
 *     a malformed body is a clean 422, never a 500.
 *   - Zero egress: only the loopback API is called (ADR-0026).
 *   - The client NEVER sends decided_tier/decided_score — the server is the
 *     sole snapshot authority (ADR-0072 D2; a stale tab must not write a
 *     stale re-entry baseline).
 *
 * Lazy `getBaseUrl()` (mirrors api/cases.ts): triageActions.ts /
 * triageDecisions.ts are transitively imported by most triage-surface
 * components and tests (TriageBanner, RecommendationCards, DashboardRoute,
 * triageBand). Resolving the base URL at call time — rather than at module
 * import time — means those tests do not all need to mock `resolveBaseUrl`/
 * `assertLoopbackBase` out of '../api/client'.
 */
import { ApiError, buildHeaders } from './client'
import type { CreateDecisionRequest, DecisionRecord, ListDecisionsResponse } from './types'

let _resolvedBase: string | null = null

function getBaseUrl(): string {
  if (_resolvedBase !== null) return _resolvedBase

  const env =
    (import.meta as { env?: { VITE_API_BASE_URL?: string; DEV?: boolean; PROD?: boolean } })
      .env ?? {}

  if (env.VITE_API_BASE_URL !== undefined) {
    _resolvedBase = env.VITE_API_BASE_URL
  } else if (env.DEV) {
    _resolvedBase = ''
  } else {
    _resolvedBase = 'http://127.0.0.1:8000'
  }

  return _resolvedBase
}

async function parseError(res: Response): Promise<ApiError> {
  const detail = await res.json().catch(() => res.text().catch(() => null))
  return new ApiError(res.status, detail, `API ${res.status}: ${res.url}`)
}

// ---------------------------------------------------------------------------
// POST /decisions — record a decision
// ---------------------------------------------------------------------------

/**
 * Record a triage decision (`expected` / `dismissed` / `false_positive`).
 *
 * The server computes `decided_tier`/`decided_score` by running the actor
 * through the pipeline (ADR-0072 D2) — never send them here.
 *
 * Returns the full record (incl. server snapshot) on 201.
 * Throws ApiError on 422 (verb/rule_name XOR mismatch) or 503 (store/pipeline
 * not wired).
 */
export async function createDecision(body: CreateDecisionRequest): Promise<DecisionRecord> {
  const res = await fetch(`${getBaseUrl()}/decisions`, {
    method: 'POST',
    headers: buildHeaders(),
    body: JSON.stringify(body),
  })
  if (!res.ok) throw await parseError(res)
  return res.json() as Promise<DecisionRecord>
}

// ---------------------------------------------------------------------------
// GET /decisions — cursor-paginated history
// ---------------------------------------------------------------------------

/**
 * Fetch the decision history (ADR-0029 D2 cursor envelope), newest-first.
 * Returns the FULL history (active + revoked) — the audit trail feeds the
 * case inbox (#16).
 *
 * `actor` scopes to one actor's full history (active + revoked).
 * Returns null on 503 (store not yet wired — honest degrade).
 */
export async function listDecisions(params?: {
  actor?: string
  cursor?: string
  limit?: number
}): Promise<ListDecisionsResponse | null> {
  const url = new URL(`${getBaseUrl()}/decisions`, globalThis.location?.origin ?? 'http://localhost')
  if (params?.actor !== undefined) url.searchParams.set('actor', params.actor)
  if (params?.cursor) url.searchParams.set('cursor', params.cursor)
  if (params?.limit !== undefined) url.searchParams.set('limit', String(params.limit))

  const res = await fetch(url.toString(), { method: 'GET', headers: buildHeaders() })
  if (res.status === 503) return null
  if (!res.ok) throw await parseError(res)
  return res.json() as Promise<ListDecisionsResponse>
}

// ---------------------------------------------------------------------------
// DELETE /decisions/{id} — soft-revoke (undo)
// ---------------------------------------------------------------------------

/**
 * Soft-revoke a decision (sets `revoked_at`) — the audit row survives
 * (append-only, ADR-0072 D2).
 *
 * Throws ApiError on 404 (unknown id) or 503 (store not wired).
 */
export async function revokeDecision(decisionId: number): Promise<void> {
  const res = await fetch(`${getBaseUrl()}/decisions/${encodeURIComponent(decisionId)}`, {
    method: 'DELETE',
    headers: buildHeaders(),
  })
  if (!res.ok) throw await parseError(res)
}

File: packages/firewatch-api/src/firewatch_api/routes/decisions.py
"""Triage-decisions routes — POST/GET /decisions, DELETE /decisions/{id}
(ADR-0072 D3, issue #47 Part 1/backend).

Routes
------
POST /decisions
    Record a triage decision (`expected` / `dismissed` / `false_positive`).
    The SERVER computes ``decided_tier``/``decided_score`` by running the
    actor through the pipeline (ADR-0072 D2) — the client never self-reports
    them. Returns 201 + the full record. 422 when ``verb='false_positive'``
    XOR ``rule_name`` is present.

GET /decisions
    Cursor-paginated list (ADR-0029 D2 envelope), newest-first. ``actor``
    scopes to one actor's full history (active + revoked).

DELETE /decisions/{id}
    Soft-revoke (sets ``revoked_at``) — the audit row survives (append-only,
    ADR-0072 D2). 404 on an unknown id.

Route class: B (writes) / C (reads) — ADR-0026 gating applies when the API
is exposed beyond loopback (loopback posture unchanged in M1, ADR-0072).

Security
--------
- All store access is parameterised SQL (adapter layer); no interpolation.
- ``actor_ip``/``rule_name``/``note`` are bounded and pattern-validated at
  the schema layer (schemas.py) — malformed input is a clean 422, never a 500.
- A ``ValueError`` raised by the store (verb/rule_name mismatch, oversized
  field) is mapped to 422 here — never leaks as an unhandled 500.
"""
from __future__ import annotations

import logging
from typing import Any

from fastapi import APIRouter, Depends, HTTPException, Query

from firewatch_api.deps import get_decision_store, get_pipeline
from firewatch_api.schemas import CreateDecisionRequest, DecisionRecord, ListDecisionsResponse
from firewatch_core.adapters.decisions.caps import DEFAULT_LIMIT, MAX_LIMIT

logger = logging.getLogger("firewatch.api.decisions")

router = APIRouter(prefix="/decisions", tags=["decisions"])


def _require_decision_store(store: Any) -> Any:
    if store is None:
        raise HTTPException(
            status_code=503,
            detail="Decision store not available — service starting up.",
        )
    return store


@router.post(
    "",
    status_code=201,
    response_model=DecisionRecord,
    summary="Record a triage decision (expected / dismissed / false_positive)",
)
async def create_decision(
    body: CreateDecisionRequest,
    store: Any = Depends(get_decision_store),
    pipeline: Any = Depends(get_pipeline),
) -> DecisionRecord:
    """Record a decision; the server computes the tier/score snapshot (ADR-0072 D2).

    Returns 503 when the decision store or pipeline is not wired.
    Returns 422 when the verb/rule_name pairing is invalid (schema layer
    already enforces the Literal verb vocabulary; this catches the XOR rule).
    """
    store = _require_decision_store(store)
    if pipeline is None:
        raise HTTPException(status_code=503, detail="Pipeline not available")

    # ADR-0072 D2 "snapshot authority is the server": run the actor through
    # the SAME pipeline every other read surface uses — never trust a
    # client-reported tier/score (a stale tab must not write a stale
    # re-entry baseline). use_ai=False: a decision snapshot is a fast,
    # deterministic fact, not an AI-augmented judgement (no LLM call here).
    score = await pipeline.analyze_ip(body.actor_ip, use_ai=False)
    decided_tier = score.escalation.tier if score.escalation is not None else None

    try:
        record = await store.create_decision(
            actor_ip=body.actor_ip,
            verb=body.verb,
            rule_name=body.rule_name,
            decided_tier=decided_tier,
            decided_score=score.score,
            note=body.note,
        )
    except ValueError as exc:
        raise HTTPException(status_code=422, detail=str(exc)) from exc
    except Exception as exc:
        logger.error("POST /decisions: create_decision failed: %s", exc, exc_info=True)
        raise HTTPException(status_code=503, detail="Decision store write failed.") from exc

    return DecisionRecord(**record)


@router.get(
    "",
    response_model=ListDecisionsResponse,
    summary="Cursor-paginated list of triage decisions",
)
async def list_decisions(
    actor: str | None = Query(
        default=None, description="Optional exact-match actor_ip filter.",
    ),
    cursor: str | None = Query(default=None, description="Opaque pagination token."),
    limit: int = Query(default=DEFAULT_LIMIT, ge=1, le=MAX_LIMIT),
    store: Any = Depends(get_decision_store),
) -> ListDecisionsResponse:
    """Return the decision history (ADR-0029 D2 envelope), newest-first.

    Returns the FULL history (active + revoked) — the audit trail feeds the
    case inbox (#16). Returns 503 when the decision store is not wired.
    """
    store = _require_decision_store(store)
    try:
        page = await store.list_decisions(limit=limit, cursor=cursor, actor=actor)
    except Exception as exc:
        logger.error("GET /decisions: list_decisions failed: %s", exc, exc_info=True)
        raise HTTPException(status_code=503, detail="Decision store read failed.") from exc
    return ListDecisionsResponse(
        items=[DecisionRecord(**item) for item in page["items"]],
        next_cursor=page["next_cursor"],
        has_more=page["has_more"],
    )


@router.delete(
    "/{decision_id}",
    status_code=200,
    response_model=None,
    summary="Soft-revoke a triage decision (undo)",
)
async def revoke_decision(
    decision_id: int,
    store: Any = Depends(get_decision_store),
) -> dict[str, Any]:
    """Soft-revoke a decision (sets ``revoked_at``) — the row is never deleted.

    Returns 404 when the decision id does not exist.
    Returns 503 when the decision store is not wired.
    """
    store = _require_decision_store(store)
    try:
        await store.revoke_decision(decision_id)
    except LookupError:
        raise HTTPException(status_code=404, detail="Decision not found.")
    except Exception as exc:
        logger.error("DELETE /decisions/%s failed: %s", decision_id, exc, exc_info=True)
        raise HTTPException(status_code=503, detail="Decision store write failed.") from exc
    return {"id": decision_id, "revoked": True}
